packages/flutter_tools/lib/src/dart/source_graph.dart

```diff
@@ -0,0 +1,257 @@
+import 'package:yaml/yaml.dart' as yaml;
+
+import 'sdk.dart';
+
+// TODO(devoncarew): Implement added, changed, and removed sources information (add a GraphChange class).
+// TODO(devoncarew): Use GraphSourceReferences.resolved when doing incremental checks.
+// TODO(devoncarew): For performance, the lexer should not lex the entire file.
+
+class SourceGraph {
+  SourceGraph(this.directory, this.entryPointPath, { this.parseEmbedderSource: false });
+
+  final Directory directory;
+  final String entryPointPath;
+  final bool parseEmbedderSource;
+
+  PackagesFile _packagesFile;
+  Map<String, String> _dartLibraryMap = <String, String>{};
+  bool _wasIncremental = false;
+  Map<String, GraphSource> _sourceMap = <String, GraphSource>{};
+
+  /// Whether the last parse was incremental, or a full re-parse.
+  bool get wasIncremental => _wasIncremental;
+
+  Iterable<GraphSource> get sources => _sourceMap.values;
+
+  void initialParse() {
+    _fullParse();
+  }
+
+  // TODO(devoncarew): Implement an incremental mode.
+  void reparseSources() {
+    _fullParse();
+  }
+
+  /// Flush any cached file contents.
+  void flushCachedFileContents() {
+    for (GraphSource source in sources)
+      source.flushCachedFileContents();
+  }
+
+  void _fullParse() {
+    _wasIncremental = false;
+
+    _sourceMap.clear();
+
+    if (parseEmbedderSource) {
+      // Parse the libraries.dart file - use that info to resolve dart: references.
+      _dartLibraryMap.clear();
+      String sdkPath = dartSdkPath;
+      File librariesFile = new File(path.join(sdkPath, 'lib/_internal/libraries.dart'));
+      if (librariesFile.existsSync()) {
+        SdkLibrariesReader reader = new SdkLibrariesReader(false);
+        LibraryMap dartLibraryMap = reader.readFromFile(
+          new JavaFile(librariesFile.path), librariesFile.readAsStringSync()
+        );
+
+        String libPath = path.normalize(path.absolute(path.join(dartSdkPath, 'lib')));
+        for (SdkLibrary library in dartLibraryMap.sdkLibraries)
+          _dartLibraryMap[library.shortName] = path.join(libPath, library.path);
+      }
+    }
+
+    // Parse the .packages file.
+    _packagesFile = new PackagesFile(new File(path.join(directory.path, '.packages')));
+    _packagesFile.parse();
+
+    GraphSource root = _getCreateSource(path.join(directory.path, entryPointPath));
+    root.parse();
+
+    // Parse any dirty sources.
+    while (sources.any((GraphSource source) => source.dirty)) {
+      for (GraphSource source in sources.where((GraphSource source) => source.dirty).toList()) {
+        source.parse();
+      }
+    }
+  }
+
+  GraphSource _getCreateSource(String resolvedPath) {
+    if (_sourceMap.containsKey(resolvedPath)) {
+      return _sourceMap[resolvedPath];
+    } else {
+      GraphSource source = new GraphSource(this, resolvedPath);
+      _sourceMap[resolvedPath] = source;
+      return source;
+    }
+  }
+
+  String _resolveDirectiveUri(String uriString, String basePath) {
```

Comment:
A minor observation: this duplicates some of the logic in SourceFactory. Possibly more importantly, it doesn't match the VM's rules for URI equality. I don't know what the VM's hot-patching API looks like, but if you're sending a mapping from URI to file content, there might be a subtle bug here related to relative URI's. We should talk off-line.